src/sqlfluff/core/parser/segments/base.py

```diff
@@ -849,6 +880,42 @@ def _preface(self, ident, tabsize) -> str:
 
     # ################ PUBLIC INSTANCE METHODS
 
+    def populate_parents(self, recurse=True) -> None:
+        """Populate parents for all segments."""
+        for seg in self.segments:
+            seg.set_parent(self)
+            # Recurse if not disabled
+            if recurse:
+                seg.populate_parents(recurse=recurse)
+
+    def set_parent(self, parent: "BaseSegment") -> None:
+        """Set the weak reference to the parent.
+
+        NOTE: Don't validate on set, because we might not have fully
+        initialised the parent yet (because we call this method during
+        the instantiation of the parent).
+        """
+        self._parent = weakref.ref(parent)
+
+    def get_parent(self) -> Optional["BaseSegment"]:
+        """Get the parent segment, with some validation.
+
+        This is provided as a performance optimisation when searching
+        through the syntax tree. Any methods which depend on this should
+        have an alternative way of assessing position, and ideally also
+        set the parent of any segments found without them.
+
+        NOTE: We only store a weak reference to the parent so it might
+        not be present. We also validate here that it's _still_ the parent
+        and potentially also return None if those checks fail.
+        """
+        if not self._parent:
+            return None
+        _parent = self._parent()
```

Comment:
maybe i'm missing something about weakref, but why is this a function call?

Reply:
That's just how `weakref` works 🤷: https://docs.python.org/3/library/weakref.html